### DISCUSSION TITLE
Bizarre experience trying to open saved SQL script file for a certain database

### DISCUSSION THREAD

**Discussion by givihuda** [dbeaver/dbeaver#16583] [Q&A] [answered]
I sometimes save scripts I put together for later use, because some tickets I work on are so complex, I can't just finish them in one session, or I need the reference later (saves me a lot of time). I've figured out how to save them to their own location on my Mac with their own file name, but DBeaver seems extremely finicky and inconsistent or buggy trying to find these files later.

For instance, if I save a .sql file for a certain connection, then try to open it during another session, sometimes/often/always DBeaver just won't show the script on the list, even though it shows many scripts from different locations (not just the obnoxious default location /Users/Me/Library/DBeaverData/workspace6/General/Scripts, which is hard to reach without special handling in Finder).

![image](https://user-images.githubusercontent.com/74785256/170105285-b5ecbcb4-d8d1-4a0f-a081-7e1733baefbf.png)

So by default DBeaver it looks like DBeaver saves the SQL query as a temporary filename in the default location, with an incremented number appended, which is completely fine, but if I 'Save As' the file to a more useful location, sometimes DBeaver just won't show the file on the list (when I right-click the DB schema and select SQL Editor -> Open SQL Script).

After having moved some of the temporary files from the default folder to their own, DBeaver would still show the files even after restart, and trying to open them from there would show an empty script tab, and then trying to open from the list again showed the file removed. There just seems to be some sort of overly-aggressive caching going on, or the list not refreshing well.

After having moved all of the files to their own folder, and restarting DBeaver, it now shows NONE on the list, and I guess DBeaver deleted all of them even though they're in their own folder, because they're gone from Finder now.

Also, why does .DS_Store show on the list?! 

This is super-confusing. There's something horribly broken about the Open SQL Script feature.

**Comment by E1izabeth** [comment, marked as answer]
DBeaver stores scripts in the project's Scripts folder, because scripts belong to the project. If you move them out of this folder or choose another project as active, they will not be shown here anymore.

If you want to add script or folder with scripts from another location to the project, you can use Link File or Link Folder option from the context menu.
![image](https://user-images.githubusercontent.com/28875055/170560510-59ef2e12-30ac-4c9c-b49b-d468cb8c9a4f.png)

**Reply by givihuda** [reply]
OK thanks. I guess I don't understand how a "project" works, as I've only used "connections". Am I not understanding the "projects" context?

I suppose you've noticed the problems I pointed out, such as a date of 1969 being displayed... something's not right about the scripts feature.

**Reply by E1izabeth** [reply]
You can learn basic usage of DBeaver projects in our video https://www.youtube.com/watch?v=Nl8GSqAUiTA . May be it would be useful for you to create a project to specify exact path to store connections and scripts.

I'm not sure how to reproduce the issue with a scripts date and why it may happen. Did you notice when (after what actions) it happened?

**Reply by givihuda** [reply]
Is using a "project" in DBeaver "basic usage"? I'm very familiar with much basic usage of DBeaver, in fact I know many things other users I work with don't about it and offer suggestions (and they may consider that intermediate or advanced usage). Project usage seems like intermediate usage, and I haven't knowingly worked with "project" in DB apps nor do I know of any coworkers who've used a feature like that in their DB apps.

The scripts date 1969 thing just showed up there, I just noticed it a while ago... I've been using the app on this computer for over a year. The whole scripts thing has been very buggy, though I'm only really noticing them now because I'm trying to clean up and organize my scripts, which is why I pointing them out recently. I can give more details if I go to clean them up again. The feature appears to not quite be done in the oven yet.

**Comment by givihuda** [comment]
@E1izabeth I'm not seeing where you're getting the context menu to add a script folder.

> DBeaver stores scripts in the project's Scripts folder, because scripts belong to the project.

I've never set up a project, yet the Scripts folder shows up in any context when I right-click a database connection or a schema in the Database Navigator pane.

![image](https://user-images.githubusercontent.com/74785256/202826348-34b97432-ffc1-4579-886f-280d2e30bede.png)

I don't see a "Project" tab anywhere. How can I not have a project, but a Scripts box showing all scripts I've ever made (as "New SQL Script")? 

> you can use Link File or Link Folder option from the context menu

I'm not seeing where you're finding this... I found a Projects tab (in Window > Projects) and the Scripts line that I can expand, but I don't get a 'Create' option when right-clicking a .sql file...

![image](https://user-images.githubusercontent.com/74785256/202826580-5a23d616-4bf4-41e3-b477-961a019958c3.png)

Honestly I'm not sure why the Database Navigator and Projects panes are almost identical... but the Database Navigator pane doesn't include the Scripts line. I suppose the Projects pane would be better overall for me to use if I just need to view/change tables and columns and properties etc, and look at scripts?

**Comment by givihuda** [comment]
@E1izabeth OK I found something else that worked... right-clicking the Scripts "folder" (in the Projects pane) and selecting Create > Link Folder.

![image](https://user-images.githubusercontent.com/74785256/202826762-8d3f2b9c-d7c8-4714-8f2f-008c30c3f6dd.png)

However, when I open a script there, it says <none> and isn't attached to any database connection or schema... which it was when it was in the default location. I merely moved the files and changed their names.

I'm not seeing any way to re-attach the script to the database connection or schema.
